rfcs/constructor-unboxing.md

```diff
@@ -0,0 +1,169 @@
+| `C3 false` 	| `2`               	|
+| `C3 true` 	| `3`               	|
+| `C4 c`    	| `4 + Char.code c`  	|
+
+More generally, we can unbox the definition `t` if:
+
+ * the *block-representations* of `t1`...`tn` are all disjoint (and distinct from `float` and from `t` for non-unary constructors)
+
+ * the *immediate-representations* of `t1`...`tn` together with `C1`...`Cn` cover less than the *immediate space*.
+
+(Here *block-representation* indicates the *set* of tags that non-immediate values of a type can have, and *immediate-representation* indicates the *number* of distinct immediate values of a given type.  *Covering less than the immediate space* means that the sum of the immediate representations of `t1`...`tn` is less than `max_int`.)
+
+### How does this relate to the existing [@@unboxed] annotation?
+
+This proposal is a conservative extension of the existing `[@@unboxed]` annotation (PRs [#606](https://github.com/ocaml/ocaml/pull/606), [#2188](https://github.com/ocaml/ocaml/pull/2188)).  It has no effect on existing code, and the extended meaning of `[@@unboxed]` is compatible with the existing meaning.
+
+### How does this relate to the existing proposal for unboxed types?
+
+[Another open proposal](https://github.com/ocaml/RFCs/pull/10) involves unboxing field types into parent types --- for example, producing flat representations of `int32` pairs:
+
+```ocaml
+type t = { x : #int32; y : #int32 }
+```
+
+The two proposals are distinct, but complementary.  The field-unboxing proposal does not support the compact rope representation, the current proposal does not support pair unboxing, and the combination of the two proposals can support representations that are more compact than either proposal supports in isolation.  In the following example, the `int32` arguments are unboxed into a single block using field unboxing, and the block associated with the `Pair` constructor is eliminated using constructor unboxing:
+
+```ocaml
+type t = { x : #int32; y : #int32 }
+type topt = Nopair | Pair of t [@@unboxed]
+```
+
+Without any support for unboxing, the value `Pair {x=0l; y=0l}` involves 4 blocks; with field unboxing alone it involves 2 blocks; with constructor unboxing alone it involves 3 blocks; with both field and constructor unboxing it involves 1 block.
+
+(There are some additional connections and distinctions between the two proposals.  The current proposal can be seen as a generalization of the *nullable types* mentioned in the field unboxing proposal.  And the field unboxing proposal is more ambitious: it additionally introduces compound unboxed types and changes to the type system to combine unboxing with abstraction.)
+
+### Extension: partial unboxing
+
+The current proposal requires all unary constructor arguments to have distinguishable representations.  It might also be useful to support the case where only some of the arguments are distinguishable by allowing per-constructor unboxing specifications.
```

Comment:
What I suggested above is already proposed here.
Actually, I find the special handling of unary constructors confusing.
All the more when you  think of types such as
```ocaml
type t = A of {x: int}
```
One could also want to flatten n-ary constructors:
```
type variance = V of bool * bool * bool [@@unboxed]
```